kyo-prelude/shared/src/main/scala/kyo/Batch.scala

```diff
@@ -108,62 +108,63 @@ object Batch:
       */
     def run[A: Flat, S, S2](v: A < (Batch[S] & S2))(using Frame): Seq[A] < (S & S2) =
 
-        case class Cont(op: Op[?, S], cont: Any => (Cont | A) < (Batch[S] & S & S2))
-        def runCont(v: (Cont | A) < (Batch[S] & S & S2)): (Cont | A) < (S & S2) =
+        type SourceAny = Source[Any, Any, S]
+        type ContAny   = Any => (ToExpand | Expanded | A) < (Batch[S] & S & S2)
+
+        case class ToExpand(op: Seq[Any], cont: ContAny)
+        case class Expanded(value: Any, source: SourceAny, cont: ContAny)
+
+        def runCont(v: (ToExpand | Expanded | A) < (Batch[S] & S & S2)): (ToExpand | Expanded | A) < (S & S2) =
             ArrowEffect.handle(erasedTag[S], v) {
-                [C] => (input, cont) => Cont(input, v => cont(v.asInstanceOf[C]))
+                [C] =>
+                    (input, cont) =>
+                        val contAny = cont.asInstanceOf[ContAny]
+                        input match
+                            case Call(v, source) => Expanded(v, source.asInstanceOf[SourceAny], contAny)
+                            case v: Seq[C]       => ToExpand(v, contAny)
             }
         end runCont
```

Comment:
This method now returns `Expanded` directly to avoid the translation from `Cont` to `Expanded` in the previous impl